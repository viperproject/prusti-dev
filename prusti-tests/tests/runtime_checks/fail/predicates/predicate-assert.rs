//@run: 101
use prusti_contracts::*;

// This test has 2 purposes:
// 1. Prusti's specification checker makes sure that predicates are not called from
// normal user code. With runtime checks this got a bit more complicated, since now
// predicates can (sometimes, after ast rewriting) be called from within user code.
// But then they're in a block starting with a #[check_only] marked closure. This test
// makes sure prusti properly recognizes this block
// 2. It also checks that the runtime check is properly executed and inserted
predicate!{
    #[insert_runtime_check]
    fn even(x: i32) -> bool {
        x % 2 == 0
    }
}

#[trusted]
fn main() {
    prusti_assert!(#[insert_runtime_check] even(3));
}
