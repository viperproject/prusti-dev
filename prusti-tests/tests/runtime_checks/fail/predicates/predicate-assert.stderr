thread 'main' panicked at $DIR/predicate-assert.rs:20:44:
Prusti Runtime Checks: Contract prusti_assert!(even(3)) was violated at runtime
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
